temporal/server.go

```diff
@@ -194,18 +198,32 @@ func (s *Server) Start() error {
 			return err
 		}
 
+		s.serviceStoppedChs[svcName] = make(chan struct{})
+
 		var svc common.Daemon
 		switch svcName {
 		case primitives.FrontendService:
 			svc, err = frontend.NewService(params)
 		case primitives.HistoryService:
-			var historySvc *history.Service
+			// todo: generalize this custom case logic as other services onboard fx
 			histApp := fx.New(
-				fx.Supply(params),
-				history.Module,
-				fx.Populate(&historySvc))
+				fx.Supply(
+					params,
+					s.serviceStoppedChs[svcName],
+				),
+				history.Module)
 			err = histApp.Err()
-			svc = historySvc
+			if err != nil {
+				close(s.serviceStoppedChs[svcName])
+				return fmt.Errorf("unable to construct service %q: %w", svcName, err)
+			}
+			s.serviceApps[svcName] = histApp
+			err = histApp.Start(ctx.Background())
```

Comment:
Do you think we need some reasonable timeout here to prevent this call from hang?